### ISSUE TITLE
prototype a deep natural selection model

### ISSUE THREAD

**Issue by matsen** [matsengrp/epam#31]
# 👉 👉 👉 👉 👉  the stuff below here is out of date. It's all implemented. see the `31-dnsm` branch

Overall we're going to want rounds of optimization in which 

1. we optimize branch lengths
2. we optimize the model

Just because NNs have a lot of parameters, we'll want to do 2 a lot more than 1.

Here's a sketch of how this could go.

### Custom Dataset

So we're going to want to store our branch length estimate for PCPs.
The nicest way to do this is to store in a torch Dataset as follows.

Note that we'll want to move everything into PyTorch tensors of indices of NTs and AAs.
Also, as per the below, BCELoss wants a float tensor even though it's an indicator.

```python
from torch.utils.data import Dataset
import torch
from epam.sequences import translate_sequences

class PCPDataset(Dataset):
    def __init__(self, nt_parents, nt_children, initial_branch_lengths):
        assert len(nt_parents) == len(nt_children), "Lengths of nt_parents and nt_children must be equal."
        assert len(nt_parents) == len(initial_branch_lengths), "Lengths of nt_parents and initial_branch_lengths must be equal."
        
        self.nt_parents = nt_parents
        self.nt_children = nt_children
        self.branch_lengths = initial_branch_lengths

        self.aa_parents = translate_sequences(self.nt_parents)
        self.aa_children = translate_sequences(self.nt_children)
        
        # Create aa_subs_indicator for amino acid substitutions
        self.aa_subs_indicator = [torch.tensor([p != c for p, c in zip(parent, child)]) for parent, child in zip(self.aa_parents, self.aa_children)]

    def __len__(self):
        return len(self.nt_parents)
    
    def __getitem__(self, idx):
        return {
            "nt_parent": self.nt_parents[idx], 
            "nt_child": self.nt_children[idx], 
            "aa_parent": self.aa_parents[idx], 
            "aa_child": self.aa_children[idx], 
            "branch_length": self.branch_lengths[idx],
            "aa_subs_indicator": self.aa_subs_indicator[idx]
        }
    
    def update_branch_lengths(self, new_estimates):
        assert len(new_estimates) == len(self.branch_lengths), "Lengths of new_estimates and existing branch_lengths must be equal."
        self.branch_lengths = new_estimates
```

We'll want to update this via the `find_optimal_branch_length` method.

### Natural selection model

We'll have a new class type called DNSM, and here's a simple (!) transformer version of this:

```python
class TransformerDNSM(nn.Module):
    def __init__(self, model_dim, head_count, layer_count):
        super(MutationPredictor, self).__init__()
        # batch_first means that we have data laid out in terms of (batch, sequence_length, features)
        self.encoder_layer = nn.TransformerEncoderLayer(model_dim, head_count, batch_first=True)
        # This just makes a stack of layer_count of the encoder_layer.
        self.encoder = nn.TransformerEncoder(self.encoder_layer, layer_count)
        self.linear = nn.Linear(model_dim, 1)

    def forward(self, src):
        out = self.encoder(src)
        out = self.linear(out)
        return torch.sigmoid(out).squeeze(-1)
```

The `squeeze` is because we are making a 1D prediction along the sequence, so the result will be the substitution probability of the column-th position for the row-th sequence in the batch.

### Incorporating into MutSel

Here's a simple implementation:

```python

class TrainableMutSel(MutSel):
    def __init__(self):
        self.selection_model = TransformerDNSM()

    def build_selection_matrix_from_parent(self, parent):
        # Whatever is needed to prep tensor for a single prediction
        parent_tensor = torch.FloatTensor(parent).unsqueeze(0)
        sel_matrix = self.selection_model(parent_tensor)
        return sel_matrix
```
Note that this is kind of crappy, because we are doing one sequence at a time.
Of course, we should be doing a batch at a time, in which case it might make sense to store `codon_probs_v` as in `build_codon_mutsel_v` in the Dataset.
We should be able to batch up the rest of the operations.

### Training

Here's how training could go.

Let's assume that we have a method `aa_subs_probs_of_parent_and_branch_length` which is batch-parallelized and makes predictions of substitution probabilities.

```python
import torch
from torch.utils.data import DataLoader
import torch.optim as optim
import torch.nn as nn

batch_size = 32
pcp_dataset = PCPDataset(nt_parents, nt_children, initial_branch_lengths)
data_loader = DataLoader(pcp_dataset, batch_size=batch_size, shuffle=True)

model = TrainableMutSel()
criterion = nn.BCELoss(reduction="mean")  # Binary cross entropy loss; mean across sites

# Only optimize the selection model
optimizer = optim.SGD(model.selection_model.parameters(), lr=0.01)

# Training loop
num_epochs = 100
for epoch in range(num_epochs):
    for batch in data_loader:
        # Get data from the current batch
        nt_parent = batch["nt_parent"]
        aa_parent = batch["aa_parent"]
        aa_child = batch["aa_child"]
        branch_length = batch["branch_length"]
        aa_subs_indicator = batch["aa_subs_indicator"].float()  # Make sure it's float for BCELoss

        # We may want to use some cached nt mutation information rather than including nt_parent here
        pred_subs_probs = model.aa_subs_probs_of_parent_and_branch_length(aa_parent, nt_parent, branch_length)

        # Compute the loss between predicted and true values
        loss = criterion(pred_subs_probs, aa_subs_indicator)

        # Zero the gradients
        optimizer.zero_grad()

        # Backward pass
        loss.backward()

        # Update parameters
        optimizer.step()

    # Output some information
    print(f"Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item():.4f}")

torch.save(model.state_dict(), "model.pth")
```

After a bunch of this sort of training, we'd re-fit the branch lengths and train the selection model again.

We have a very imbalanced setup with many more 0s than 1s. 
We could set a weight to correct this imbalance with [BCEWithLogitsLoss](https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html).
It's also more numerically stable because incorporates a final sigmoid layer in the loss (**so you don't want to have that final layer in the nn**).
